speculation-rules.bs

```diff
@@ -231,7 +276,63 @@ Inside the [=prepare the script element=] algorithm we make the following change
   1. If |input|["`target_hint`"] [=map/exists=]:
     1. If |input|["`target_hint`"] is not a [=valid browsing context name or keyword=], then return null.
     1. Set |targetHint| to |input|["`target_hint`"].
-  1. Return a [=speculation rule=] with [=speculation rule/URLs=] |urls|, [=speculation rule/requirements=] |requirements|, and [=speculation rule/target browsing context name hint=] |targetHint|.
+  1. Return a [=speculation rule=] with [=speculation rule/URLs=] |urls|, [=speculation rule/predicate=] |predicate|, [=speculation rule/requirements=] |requirements|, and [=speculation rule/target browsing context name hint=] |targetHint|.
+</div>
+
+<div algorithm="parse a document rule predicate">
+  To <dfn>parse a document rule predicate</dfn> given a value |input| and [=URL=] |baseURL|:
+
+  1. If |input| is not a [=map=], then return null.
+  1. If |input| does not [=map/contain=] exactly one of "`and`", "`or`", "`not`", "`href_matches`" and "`selector_matches`", then return null. Otherwise, let |predicateType| be that key.
+  1. If |predicateType| is "`and`" or "`or`", then:
+    1. If |input| has any [=map/key=] other than |predicateType|, then return null.
+    1. Let |rawClauses| be the |input|[|predicateType|].
+    1. If |rawClauses| is not a [=list=], then return null.
+    1. Let |clauses| be an empty [=list=].
+    1. [=list/For each=] |rawClause| of |rawClauses|:
+      1. Let |clause| be the result of [=parsing a document rule predicate=] given |rawClause|.
+      1. If |clause| is null, then return null.
+      1. [=list/Append=] |clause| to |clauses|.
+    1. If |predicateType| is "`and`", then return a [=document rule conjunction=] whose [=document rule conjunction/clauses=] is |clauses|.
+    1. If |predicateType| is "`or`", then return a [=document rule disjunction=] whose [=document rule disjunction/clauses=] is |clauses|.
+  1. If |predicateType| is "`not`", then:
+    1. If |input| has any [=map/key=] other than "`not`", then return null.
+    1. Let |rawClause| be the |input|[|predicateType|].
+    1. Let |clause| be the result of [=parsing a document rule predicate=] given |rawClause|.
+    1. If |clause| is null, then return null.
+    1. Return a [=document rule negation=] whose [=document rule negation/clause=] is |clause|.
+  1. If |predicateType| is "`href_matches`":
+    1. If |input| has any [=map/key=] other than "`href_matches`", then return null.
+    1. Let |rawPatterns| be |input|["`href_matches`"].
+    1. If |rawPatterns| is not a [=list=], then set |rawPatterns| to « |rawPatterns| ».
+    1. Let |patterns| be an empty [=list=].
+    1. [=list/For each=] |rawPattern| of |rawPatterns|:
```

Comment:
So this allows `"href_matches": ["pattern1", "pattern2"]` instead of requiring the more explicit `"or": [{"href_matches": "pattern1"}, {"href_matches": "pattern2"}]`. I'm not sure that's a good tradeoff? Especially for selectors where we now have *three* ways of doing "or", the third being `"selector_matches": "selector1, selector2"`?

Reply:
Your versus example is tough... the latter version is clearly more concise, but it's still not obvious to me that it's an "or" operation. Hmm. I guess I'm OK keeping it.

Reply:
Acknowledged.